data/src/main/java/com/linkedin/util/LineColumnNumberWriter.java

```diff
@@ -0,0 +1,193 @@
+package com.linkedin.util;
+
+import java.io.IOException;
+import java.io.Writer;
+import java.util.Objects;
+import java.util.Stack;
+import java.util.function.Predicate;
+
+
+/**
+ * Wraps a {@link Writer} and tracks current line and column numbers
+ */
+public class LineColumnNumberWriter extends Writer {
+
+  private final Writer _writer;
+  private final Stack<CharacterPosition> _savedPositionStack = new Stack<>();
+  private int _column;
+  private int _line;
+  private int _previousChar;
+  private Predicate<Character> _isWhitespaceFunction;
+  private final CharacterPosition _lastNonWhitespacePosition;
+
+  /**
+   * Creates a new writer.
+   *
+   * @param out a Writer object to provide the underlying stream.
+   */
+  public LineColumnNumberWriter(Writer out) {
+    _writer = out;
+    _column = 1;
+    _line = 1;
+    _previousChar = -1;
+    _isWhitespaceFunction = (Character::isWhitespace);
+    _lastNonWhitespacePosition = new CharacterPosition(0, 0);
+  }
+
+  /**
+   * Returns 1 based indices of row and column next character will be written to
+   */
+  public CharacterPosition getCurrentPosition() {
+    return new CharacterPosition(_line, _column);
+  }
+
+  /**
+   * Returns 1 based indices of last row and column ignoring trailing whitespace characters
+   */
+  public CharacterPosition getLastNonWhitespacePosition() {
+    return _lastNonWhitespacePosition;
+  }
+
+  /**
+   * Saves current row and column to be retrieved later by calling {@link #popSavedPosition()}
+   *
+   * Saved positions are stored in a stack so that calls to saveCurrentPosition() and
+   * {@link #popSavedPosition()} can be nested. Saved positions are adjusted to skip whitespace to make it
+   * easier to get actual token start positions in indented output. If you call saveCurrentPosition() at column x
+   * and then write four spaces followed by non-whitespace, the column number returned by
+   * {@link #popSavedPosition()} will be x + 4.
+   */
+  public void saveCurrentPosition() {
+    _savedPositionStack.push(new CharacterPosition(_line, _column));
+  }
+
+  /**
+   * Retrieves row and column from the last time {@link #saveCurrentPosition()} was called
+   */
+  public CharacterPosition popSavedPosition() {
+    return _savedPositionStack.pop();
+  }
+
+  /**
+   * Override definition of whitespace used to adjust character positions to skip
+   * whitespace. By default, the definition of whitespace is provided by {@link java.lang.Character#isWhitespace}
+   */
+  public void setIsWhitespaceFunction(Predicate<Character> isWhitespaceFunction) {
+    _isWhitespaceFunction = isWhitespaceFunction;
+  }
+
+  @Override
+  public void write(char[] cbuf, int off, int len) throws IOException {
+    _writer.write(cbuf, off, len);
+    for (; len > 0; len--) {
+      char c = cbuf[off++];
```

Comment:
Is boundary check needed, or is it taken care by line 81 already?

Reply:
This is taken care of by line 81. It will throw IndexOutOfBoundsException if len/offset are invalid.